File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auton/AutonMethods.java
package org.firstinspires.ftc.teamcode.auton;

import static org.firstinspires.ftc.teamcode.utilities.CONFIG.clawRotator;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.hardware.Servo;


import org.firstinspires.ftc.teamcode.utilities.Claw;
import org.firstinspires.ftc.teamcode.utilities.Slides;
import org.firstinspires.ftc.teamcode.utilities.SimpleMecanumDrive;
import org.firstinspires.ftc.teamcode.utilities.ClawRotator;
import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.firstinspires.ftc.teamcode.utilities.Latch;

public abstract class AutonMethods extends LinearOpMode {
    public SimpleMecanumDrive drivetrain;
    public ClawRotator clawrotator;
    public Claw claw;
    public Slides slides;
    public Latch latch;

    @Override
    public void runOpMode() throws InterruptedException {

        telemetry.addData("Status", "Initializing hardware...");
        telemetry.update();

        // Initialize components and check for null references
        drivetrain = new SimpleMecanumDrive(hardwareMap);
        clawrotator = new ClawRotator(hardwareMap, telemetry);
        slides = new Slides(hardwareMap, telemetry);
        claw = new Claw(hardwareMap, telemetry);
        latch = new Latch(hardwareMap, telemetry);

        telemetry.addData("Status", "Initialization Complete");
        telemetry.update();
        waitForStart();
    }
    public final double POWER = 0.5;

    public void stopDrive() {
        drivetrain.move(0, 0, 0);
        telemetry.update();
        sleep(100);
    }

    // latch
    public void latch() {
        latch.latchOn();
    }

    public void unlatch() {
        latch.latchOn();
    }


    // claw
    public void clawOpen(){
        claw.open();
        sleep(1000);
    }

    public void clawClose() {
        claw.close();
    }

    // slides

    public void slidesExtend(){
        slides.extend();
        sleep(2000);
        slides.stop();
    }

    public void slidesRetract(){
        slides.retract();
        sleep(2000);
        slides.stop();
    }

    // Going forward, backward, turning, going left, going right

    public void moveForward(double x){
        double duration = 900 * x;
        drivetrain.move(0, POWER, 0);
        sleep((int)duration);
        stopDrive();
    }

     public void moveBackward(double x){
         double duration = 900 * x;
         drivetrain.move(0, -POWER, 0);
         sleep((int)duration);
         stopDrive();
     }

     public void moveRight(double x){
         double duration = 1500 * x;
         drivetrain.move(-POWER, 0, 0);
         sleep((int)duration);
         stopDrive();
     }

     public void moveLeft(double x){
        double duration = 1500 * x;
        drivetrain.move(POWER, 0, 0);
        sleep((int)duration);
        stopDrive();
     }

     public void rotateCw(double x){
        double duration = 3250 * x;
        drivetrain.move(0, 0, POWER);
        sleep((int)duration);
        stopDrive();
     }

     public void rotateCcw(double x){
        double duration = 3250 * x;
        drivetrain.move(0, 0, -POWER);
        sleep((int)duration);
        stopDrive();
     }

     // slides rotate

     public void slidesRotateUp() {
         slides.rotateRight();
         sleep(800);
         slides.stopRotator();
//         Runnable slidesRunnable = new Runnable() {
//             @Override
//             public void run() {
//                 sleep(10);
//                 slides.stopRotator();
//             }
//         };
//         Thread newThread = new Thread(slidesRunnable);
//         newThread.start();

     }

     public void slidesRotateDown() {
         slides.rotateLeft();
         sleep(1000);
         slides.stopRotator();
     }

     // claw drop/pick up

     public void clawDrop () {
         clawrotator.toDrop();
     }

     public void clawPick() {
         clawrotator.toPick();
     }
 }

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/auton/AutonMethods.kt
package org.firstinspires.ftc.teamcode.auton

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.teamcode.utilities.Claw
import org.firstinspires.ftc.teamcode.utilities.ClawRotator
import org.firstinspires.ftc.teamcode.utilities.SimpleMecanumDrive
import org.firstinspires.ftc.teamcode.utilities.Slides

abstract class AutonMethods : LinearOpMode() {
    var drivetrain: SimpleMecanumDrive? = null
    var clawrotator: ClawRotator? = null
    var claw: Claw? = null
    var slides: Slides? = null

    @Throws(InterruptedException::class)
    override fun runOpMode() {
        telemetry.addData("Status", "Initializing hardware...")
        telemetry.update()

        // Initialize components and check for null references
        drivetrain = SimpleMecanumDrive(hardwareMap)
        if (drivetrain == null) {
            telemetry.addData("Error", "drivetrain failed to initialize!")
        } else {
            telemetry.addData("SimpleMecanumDrive", "Initialized successfully")
        }

        clawrotator = ClawRotator(hardwareMap, telemetry)
        if (clawrotator == null) {
            telemetry.addData("Error", "ClawRotator failed to initialize!")
        } else {
            telemetry.addData("ClawRotator", "Initialized successfully")
        }

        slides = Slides(hardwareMap, telemetry)
        if (slides == null) {
            telemetry.addData("Error", "Slides failed to initialize!")
        } else {
            telemetry.addData("Slides", "Initialized successfully")
        }

        telemetry.addData("Status", "Initialization Complete")
        telemetry.update()
        waitForStart()
    }

    val POWER: Double = 2.0

    fun clawOpen() {
        claw!!.open()
        sleep(5000)
    }

    fun slidesExtend() {
        slides!!.extend()
        telemetry.update()
        sleep(6000)
    }

    fun slidesRetract() {
        slides!!.retract()
        telemetry.update()
    }

    // public void clawClose(){
    // claw.close();
    // sleep(100);
    // }
    fun stopDrive() {
        drivetrain!!.move(0.0, 0.0, 0.0)
        telemetry.update()
        sleep(100)
    }

    // Going forward, backward, turning, going left, going right
    fun backOneee() {
        //goes in reverse a little more than 1, about 1.2

        drivetrain!!.move(0.0, POWER, 0.0)
        sleep(1450)
        stopDrive()
    }

    fun rightOneHalf() {
        drivetrain!!.move(-POWER, 0.0, 0.0)
        telemetry.update()
        sleep(3000)
        stopDrive()
    }

    fun slidesRotateUp() {
        slides!!.rotateRight()
        telemetry.update()
        sleep(1000)
    }

    fun slidesRotateDown() {
        slides!!.rotateLeft()
        telemetry.update()
        sleep(1000)
    }

    fun forwardOne() {
        if (drivetrain != null) {
            drivetrain!!.move(0.0, POWER, 0.0)
            telemetry.addData("Action", "Moving forward")
        } else {
            telemetry.addData("Error", "Drivetrain is not initialized in forwardOneeee!")
        }
        telemetry.update()
        sleep(2000)
        stopDrive()
    }

    fun clawDrop() {
        clawrotator!!.toDrop()
        telemetry.update()
        sleep(1450)
    }

    fun clawPick() {
        clawrotator!!.toPick()
        telemetry.update()
    }


    fun rightLong() {
        // goes to the ending spot in auton
        drivetrain!!.move(-POWER, 0.0, 0.0)
        telemetry.update()
        sleep(8000)
        stopDrive()
    }

    fun leftOneHalf() {
        //should be a little less than 0.5 around 0.2
        drivetrain!!.move(POWER, 0.0, 0.0)
        telemetry.update()
        sleep(3000)
        stopDrive()
    }

    fun forwardHalf() {
        drivetrain!!.move(0.0, -POWER, 0.0)
        telemetry.update()
        sleep(900)
        stopDrive()
    }

    fun backOneHalf() {
        drivetrain!!.move(0.0, POWER, 0.0)
        telemetry.update()
        sleep(450)
        stopDrive()
    }
}



